pkg/pgmodel/pgx.go

```diff
@@ -404,76 +435,208 @@ func (p *pgxInserter) getMetricTableName(metric string) (string, error) {
 	return tableName, err
 }
 
-func (p *pgxInserter) getMetricTableInserter(metric string) chan insertDataRequest {
-	var inserterI interface{}
-	var ok bool
-	var inserter chan insertDataRequest
-	if inserterI, ok = p.metricTableInserters.Load(metric); !ok {
-		ch := make(chan insertDataRequest, 1000)
-		ii, loaded := p.metricTableInserters.LoadOrStore(metric, ch)
-		inserterI = ii
-		if !loaded {
-			go p.metricTableInsert(metric, ch)
+func (p *pgxInserter) getMetricTableInserter(metricTable string) chan insertDataRequest {
+	h := maphash.Hash{}
+	h.SetSeed(p.seed)
+	h.WriteString(metricTable)
+	if len(p.inserters) < 1 {
+		panic(fmt.Sprintf("invalid len %d", len(p.inserters)))
+	}
+	inserter := h.Sum64() % uint64(len(p.inserters))
+	return p.inserters[inserter]
+}
+
+type InsertHandler struct {
+	conn    pgxConn
+	input   chan insertDataRequest
+	pending map[string]*list.Element
+	timer   list.List // list of PendingBuffer
+
+	oldBuffers []*PendingBuffer
+}
+
+type PendingBuffer struct {
+	metricTable   string
+	needsResponse []insertDataTask
+	batch         SampleInfoIterator
+	start         time.Time
+}
+
+func runInserterRoutine(conn pgxConn, input chan insertDataRequest) {
+	handler := InsertHandler{
+		conn:    conn,
+		input:   input,
+		pending: make(map[string]*list.Element),
+		timer:   list.List{},
+	}
+
+	for {
+		if !handler.hasPendingReqs() {
+			stillAlive := handler.waitForReq()
+			if !stillAlive {
+				return
+			}
+			continue
+		}
+		receivingNewReqs := true
+		for receivingNewReqs {
+			handler.flushTimedOutReqs()
+			receivingNewReqs = handler.handleMoreReqs()
+			//TODO if we run out of new reqs, flush one?
 		}
+		handler.flushEarliestReq()
+	}
+}
+
+func (h *InsertHandler) hasPendingReqs() bool {
+	return len(h.pending) > 0
+}
 
+func (h *InsertHandler) waitForReq() bool {
+	req, ok := <-h.input
+	if !ok {
+		return false
 	}
-	inserter = inserterI.(chan insertDataRequest)
 
-	return inserter
+	h.handleReq(req)
+
+	return true
 }
 
-func (p *pgxInserter) metricTableInsert(metric string, input chan insertDataRequest) {
-	tableName, err := p.getMetricTableName(metric)
-	if err != nil {
+func (h *InsertHandler) handleMoreReqs() bool {
+	select {
```

Comment:
I wonder if this should actually be a for {select{}} with one of the cases being a timer that checks for the livelock timeout

Reply:
select performance is wonky, and this will ~never be taken; I think we should optimize for performance here